Ft signin fullstack 157912677

LuffyKing opened the pull request:
#### What does this PR do?
It sets up the client branch and implements login
#### Description of Task to be completed?
Creation of a client folder for the full stack version of TrackerHero. Creation of a login feature for the application that integrates the front end login page and the login API, this will allow users to log in to TrackerHero. Creation of a route and fetch request to protect pages that require a logged in user.
#### Any background context you want to provide?

#### What are the relevant Github Issues?
The only issue so is only the pending review
Questions:

benfluleck commented:
Why are the 2 commits having the same title? Please fix

benfluleck commented on client/UI/css/AdminViewRequests.css:
```diff
@@ -0,0 +1,78 @@
+.modalDelResolve,.approvebtn, .modalDeldisapprove,.modalCancel,
+.modalDelApprove,.disapprovebtn, .resolvebtn,.placeholder,
```
This is not proper css naming convention, please fix

benfluleck commented on client/UI/html/AdminUserRequestDetailPage.html:
```diff
@@ -0,0 +1,111 @@
+      <a class="default-primary-color text-primary-color Logo pacifico" href="SigninPage.html">TrackerHero</a>
+      <ul class="default-primary-color text-primary-color">
+        <li><a href="MakeRequest.html">Make Request &nbsp<i class="fas fa-bullhorn"></i></a></li>
+        <li><a href="#">Profile &nbsp<i class="fas fa-user"></i></a></li>
```
please use padding and not nbsp

benfluleck commented on client/UI/html/SigninPage.html:
```diff
@@ -0,0 +1,75 @@
+      <a class="default-primary-color text-primary-color Logo pacifico" href="SigninPage.html">TrackerHero</a>
+      <ul class="default-primary-color text-primary-color">
+        <li><a href="MakeRequest.html">Make Request &nbsp<i class="fas fa-bullhorn"></i></a></li>
+        <li><a href="#">Profile &nbsp<i class="fas fa-user"></i></a></li>
```
Please change all instances where you use nbsp

benfluleck commented on client/UI/js/alertAdminViewRequests.js:
```diff
@@ -0,0 +1,122 @@
+      password: form.elements.password.value
+    })
+  })
+    .then(response => response.json()).then((json) => {
```
Maybe a different name than json would be more appropriate here

benfluleck commented on client/server/validation/DatabaseHelpers.js:
```diff
@@ -0,0 +1,67 @@
+        return response.status(500).send({ message: error1.stack });
+      }
+      if (requestRow.rows.length > 0) {
+        const requestPluSing = requestRow.rows.length === 1 ? 'request has' : 'requests have';
```
what is requestPluSing?

LuffyKing replied:
The text changes based on the number of requests, If it it plural it becomes 'requests have'.

benfluleck replied:
and you decided the best name for such a variable was a name nobody could understand. Please refactor

benfluleck commented on client/server/validation/DatabaseHelpers.js:
```diff
@@ -0,0 +1,67 @@
+      }
+      if (requestRow.rows.length > 0) {
+        const requestPluSing = requestRow.rows.length === 1 ? 'request has' : 'requests have';
+        if (operation === 'get multiple requests') {
```
in this case statement would implement this a lot quicker but also why are we tightly coupling the operation here using a numeric value with a case statement would be the ideal scenario here.

benfluleck commented on client/server/validation/RejectStatusValidator.js:
```diff
@@ -0,0 +1,13 @@
+import { statusChanger } from './DatabaseHelpers';
+/**
+* It sets the status to Rejected
+* @param {Object} request - request object containing params and body
```
why are the param in capital go and check the difference between object and  Object in JS

benfluleck commented on client/server/validation/createARequestValidator.js:
```diff
@@ -0,0 +1,174 @@
+/**
+* It finds the fields that are undefined or null
+* @param {Object} reqBody - object containing the relevant field values
+* @returns {string[]} - an array of the undefined fields as strings
```
I am pretty sure this is not a type in javascript, please adjust

LuffyKing replied:
That says the input is a string of arrays

benfluleck replied:
So this is actually not a type in javascript but it is in typescript. Take note also refer to types in javascript.

benfluleck commented on client/server/validation/createARequestValidator.js:
```diff
@@ -0,0 +1,174 @@
+* It finds the fields that are supposed to have a string as a value but
+* possesses any data type instead
+* @param {number} statusCode - status code
+* @param {Object} response - response object that conveys the result of the request
```
Sometimes small letter object sometime Object, which one you dey?

benfluleck commented on client/server/validation/createARequestValidator.js:
```diff
@@ -0,0 +1,174 @@
+* possesses any data type instead
+* @param {number} statusCode - status code
+* @param {Object} response - response object that conveys the result of the request
+* @param {string[]} invalidFieldsArray - an array of invalid fields
```
see above

benfluleck commented on client/server/validation/createARequestValidator.js:
```diff
@@ -0,0 +1,174 @@
+* It validates the fields and renders the appropriate response
+* @param {Object} request - request object containing params and body
+* @param {Object} response - response object that conveys the result of the request
+* @param{Object} next - middleware that calls the net middleware in the stack
```
What is going on here please add the space

LuffyKing replied:
lol

LuffyKing committed:
Feat(login): implements fullstack login
- Setups the client server in the client folder
- Setups the client UI templates in the client folder
- Implements fetch request for login and switching to the UserViewRequests page
- Implements fetch request for protecting routes
- Adds a new route for verifying UI tokens that have been stored on localStorage
- Modified package.json to transpile client folder
- Refactored class names to BEM
- Refactored nav spacing to use padding
- Refactored jsdoc comments
- Refactored code to meet LF Benny's requirements
[Delivers #157912677]
